File: backend/app/services/paper_service.py
import httpx

BASE_URL = "https://api.openalex.org/works"


def reconstruct_abstract(inverted_index):
    if not inverted_index:
        return ""
    word_positions = []
    for word, positions in inverted_index.items():
        for pos in positions:
            word_positions.append((pos, word))
    word_positions.sort(key=lambda x: x[0])
    return " ".join(word for _, word in word_positions)


def search_papers(query: str):

    params = {
        "search": query,
        "per-page": 10
    }

    response = httpx.get(
        BASE_URL,
        params=params
    )

    data = response.json()

    papers = []

    for item in data.get("results", []):

        authors = []

        for author in item.get("authorships", []):
            if "author" in author and "display_name" in author["author"]:
                authors.append(author["author"]["display_name"])

        abstract = reconstruct_abstract(item.get("abstract_inverted_index"))

        papers.append(
            {
                "title": item.get("title") or "Untitled Paper",
                "year": item.get("publication_year"),
                "citations": item.get("cited_by_count", 0),
                "authors": authors,
                "url": item.get("doi") or item.get("id"),
                "abstract": abstract
            }
        )

    return papers

File: frontend/src/services/paperService.js
import axios from "axios";

const API_URL = import.meta.env.VITE_API_URL || "https://research-paper-explorer-751e.onrender.com";


export async function searchPapers(query) {
    const response = await axios.get(
        `${API_URL}/papers/search`,
        {
            params: {
                query: query
            }
        }
    );

    return response.data;
}

export async function analyzePaper(title, abstract) {
    const response = await axios.post(
        `${API_URL}/ai/analyze`,
        {
            title: title,
            abstract: abstract || "No abstract provided."
        }
    );

    return response.data;
}